sympy/geometry/tests/test_plane.py

```diff
@@ -93,7 +95,7 @@ def test_plane():
     assert pl6.distance(pl6.p1) == 0
     assert pl7.distance(pl6) == 0
     assert pl7.distance(l1) == 0
-    assert pl6.distance(Segment3D(Point3D(2, 3, 1), Point3D(1, 3, 4))) == 0
+    assert pl6.distance(Segment3D(Point3D(2, 3, 1), Point3D(1, 3, 4))) == 4*sqrt(3)/3
```

Comment:
This test [fails in Travis](https://travis-ci.org/sympy/sympy/jobs/421226250#L2061-L2069) for some reason. If I run it manually with pytest on my system, the precise issue is that the left-hand side returns ``2*sqrt(3)`` instead of ``4*sqrt(3)/3``. Both values are distances between the endpoints of the segment and ``pl6`` (which are the only possible distances given that the segment and ``pl6`` do not intersect), but ``4*sqrt(3)/3`` is clearly smaller so it should be the correct value.
@czgdp1807 Does it work fine on your system?
Otherwise I guess it's a bug in ``Plane.distance``?

Reply:
On my system it gives 4*sqrt(3)/3. My SymPy version is 1.1.1.

Reply:
@czgdp1807 That's pretty weird then.
Regarding the SymPy version, aren't you testing it directly on top of the repository (e.g. installed via ``pip -e /path/to/your/git/repo``)?

Reply:
Thanks for giving this information. I will do this from next time.

Reply:
It's returning the maximum distance. I ran the following piece of code,

```
from sympy import *
pl6 = Plane(Point3D(2, 3, 7), normal_vector=(2, 2, 2))
print(pl6.distance(Segment3D(Point3D(2, 3, 1), Point3D(1, 3, 4))))
print(pl6.distance(Point3D(1, 3, 4)))
print(pl6.distance(Point3D(2, 3, 1)))
```
I got the following output,
```
2*sqrt(3)
4*sqrt(3)/3
2*sqrt(3)
```
So which to keep?
According to me, returning the distance of a non-intersecting, non-parallel line to a plane isn't intuitive nor correct. However, we should return an unevaluated object. What do you think? @smheidrich

Reply:
@sidhantnagpal Any comments/reviews from your side?

Reply:
@oscarbenjamin @oscargus Can you please help me with this issue?

Reply:
I don't know the geometry package well. Is this the code responsibly for this:
[https://github.com/sympy/sympy/blob/4063057bbe136d1a2a1d4a6b325565259eadeb84/sympy/geometry/plane.py#L292](https://github.com/sympy/sympy/blob/4063057bbe136d1a2a1d4a6b325565259eadeb84/sympy/geometry/plane.py#L292)

Reply:
Yes @oscarbenjamin . The code uses the statements written under that if-branch. I basically want to know your view, on my [third last comment](https://github.com/sympy/sympy/pull/15102/files#r270671857). What does your intuition say? Should we return the maximum distance or minimum distance between a non-intersecting line and plane pair?

Reply:
I would have thought that closest distance makes sense. What do the distance methods of other objects do in similar situations?

Reply:
The distance between line and a point comes out to be correct, since there is no ambiguity.
I think you are right because, the distance between a line and a point is defined as the perpendicular distance of the point from the line. However, since the perpendicular distance is the minimum distance that the point is having from the line. So, we should follow the same convention in 3D, that is return the minimum distance. I think the current code needs correction in the regard. 
Do you know someone well versed with this module, so that he/she can provide the logic behind returning the maximum distance?

Reply:
> The distance between line and a point comes out to be correct, since there is no ambiguity.

Then it isn't a good example to check. What do other examples do?

Reply:
I could think of the following cases for distance pairs:

1. Plane and Point
2. Plane and Line(which is to be corrected)
3. Line and Line
4. Point and Line
5. Plane and Plane
6. Polygon and Point
7. Polygon and Line

The closest example which I can see in the above cases is the 7th which is Polygon and Line. However, it is not implemented yet in SymPy. So, the second most closest is 4th, which I have described in my previous comment.
(1, 3, 4, 5, 6) are quite unambiguous cases and work as expected in SymPy.

Reply:
I tried the following piece of code.

```
from sympy import *
plane1 = Plane(Point3D(2, 3, 7), normal_vector=(2, 2, 2))
plane2 = Plane(Point3D(2, 3, 7), normal_vector=(3, 3, 3))
point = Point3D(1, 3, 4)
line1 = Segment3D(Point3D(2, 3, 1), Point3D(1, 3, 4))
line2 = Segment3D(Point3D(7, 8, 9), Point3D(5, 6, 7))
p1, p2, p3, p4 = map(Point, [(0, 0), (1, 0), (5, 1), (0, 1)])
poly1 = Polygon(p1, p2, p3, p4)
try:
    print(plane1.distance(plane2))
except Exception as e:
    print("Error-", str(e))
try:
    print(line1.distance(line2))
except Exception as e:
    print("Error-", str(e))
try:
    print(plane1.distance(plane2))
except Exception as e:
    print("Error-", str(e))
try:
    print(poly1.distance(Point(2, 3)))
except Exception as e:
    print("Error-", str(e))
try:
    print(poly1.distance(line1))
except Exception as e:
    print("Error-", str(e))
```

Output:
```
0
Error- 
0
2
Error- 
```
I also found that we can have polygons with objects of Point3D class.

Reply:
Does Line and Line give the distance between the closest points on each line? Does the documentation anywhere say what distance should return?

Reply:
If the Lines are infinite, then there will be no ambiguity. If the lines are bounded that is, objects of `Segment` or `Segment3D` then, the [this](https://github.com/sympy/sympy/blob/master/sympy/geometry/line.py#L1612) suggests that it can only calculate the shortest distance between the segment and line. Hence, this strongly suggests that, keeping minimum distance is a good and consistent convention.

Reply:
@oscarbenjamin @sidhantnagpal Please provide your views on [this](https://github.com/sympy/sympy/pull/15102#discussion_r272785546)

Reply:
The docstring of `Segment.distance` expects the other argument to be a `Point` (and it raises `NotImplementedError` in all other cases). Besides, the method under consideration is `Plane.distance`. I agree that the minimum distance should be returned for such cases, however for `Segment` and `Ray` it is not as trivial as evaluating at the extremities. It requires obtaining the closest point on it with respect to the plane (which can be anywhere on the LinearEntity).

I have included the fix here based on https://github.com/sympy/sympy/pull/15102#discussion_r212788516 with additional improvements in #16637. I have included the required changes for `Segment3D` and `Ray3D` based on parametric form in #16637.

Reply:
For the polygon-line case, one way could be to find the distance of the line from all the polygon edges (which are nothing but segments) and return the minimum distance obtained.

There has been some discussion for the polygon-polygon case in https://github.com/sympy/sympy/issues/5216.